### PR TITLE
Change NameNode to a struct

### PR DESCRIPTION
**Pull request by Shane32**
NameNode simply contains a string and a SourceLocation.  Since Name is not accessible via interfaces, we can be sure it is not boxed when the Name is being read.  Most of the classes that use NameNode also provide a property like Name that directly retrieves the string from the NameNode.  There are certainly times when boxing can occur, but it seems to be pretty rare.  Seems like a good candidate for a struct.  I have not yet run any performance tests.

Note that it's even possible that SourceLocation is never read and could be removed from the struct.  This would take more investigation.

### CODE REVIEW

**Comment by Shane32** [comment]
If you like where this is heading, let's wait until SourceLocation is a struct before we fully review this PR.

File: docs2/site/docs/migrations/migration4.md [hunk 1]
```diff
@@ -20,6 +20,7 @@
 * `IHaveDefaultValue.Type` has been moved to `IProvideResolvedType.Type`
 * `Connection<TNode, TEdge>.TotalCount` has been changed from an `int` to an `int?`. This allows for returning `null` indicating that the total count is unknown.
 * By default fields returned by introspection query are no longer sorted by their names. `LegacyV3SchemaComparer` can be used to switch to the old behavior.
+* `NameNode` has been changed to a struct.
```
Comment (sungam3r) [inline docs2/site/docs/migrations/migration4.md:23]:
```suggestion
* `NameNode` has been changed to a readonly struct.
```

**Comment by sungam3r** [comment]
Let's merge #2116 first.

**Comment by sungam3r** [comment]
Waiting for #2111

**Comment by Shane32** [comment]
I seriously don't know when the labeler is useful.  Every time I make a change, I have to undo stuff like this:

image

**Comment by Shane32** [comment]
We most certainly don't need labels to tell us when the tests change.  They should change with every bugfix.  Documentation labels should label when we are updating documentation, which could be code comments.  The labeler doesn't know when I do that.  It really is pointless.

**Comment by sungam3r** [comment]
Current rules:
```
test:
  - src/GraphQL.DataLoader.Tests/**/*
  - src/GraphQL.Dummy/**/*
  - src/GraphQL.Harness.Tests/**/*
  - src/GraphQL.Tests/**/*

CI:
  - .github/workflows/**/*

code style:
  - .editorconfig

documentation:
  - README.md
  - docs2/**/*

performance:
  - src/GraphQL.Benchmarks/**/*
```

Change rules as you see fit.

**Comment by sungam3r** [comment]
#2111 merged

**Comment by Shane32** [comment]
Another thing we could consider is removing NameNode entirely.

File: src/GraphQL/Language/AST/NameNode.cs [hunk 2]
```diff
@@ -1,21 +1,39 @@
...
             Name = name;
+            SourceLocation = default;
         }
 
         /// <summary>
         /// Returns the contained name.
         /// </summary>
         public string Name { get; }
+
+        IEnumerable<INode> INode.Children => throw new NotImplementedException();
```
Comment (sungam3r) [inline src/GraphQL/Language/AST/NameNode.cs:34]:
```suggestion
        IEnumerable<INode> INode.Children => null;
```

To align behavior with other leaf nodes.

**Comment by sungam3r** [comment]
> Another thing we could consider is removing NameNode entirely.

I'm not sure.

File: src/GraphQL/Language/AST/NameNode.cs (hunk 2 above)
Comment (sungam3r) [reply src/GraphQL/Language/AST/NameNode.cs:34]: It's still safer.

File: src/GraphQL/Language/AST/NameNode.cs [hunk 3]
```diff
@@ -1,21 +1,39 @@
+using System;
```
Comment (sungam3r) [inline src/GraphQL/Language/AST/NameNode.cs:1]:
```suggestion
```

**Review by sungam3r — APPROVED** [review]

**Pull request merged**